packages/big-design/src/components/Tree/TreeNode/TreeNode.tsx

```diff
@@ -0,0 +1,328 @@
+import { CheckIcon, ChevronRightIcon, FolderIcon } from '@bigcommerce/big-design-icons';
+import React, { Dispatch, useCallback, useEffect, useMemo, useRef, useState } from 'react';
+
+import { typedMemo } from '../../../utils';
+import { StyledCheckbox } from '../../Checkbox/private';
+import { FlexItemProps } from '../../Flex';
+import { StyledRadio } from '../../Radio/styled';
+import { DelayedSpinner } from '../DelayedSpinner';
+import { useIsExpanded, useIsSelected } from '../hooks';
+import { Action } from '../reducer';
+import { StyledUl } from '../styled';
+import { TreeNodeProps, TreeNodeRef, TreeProps, TreeState } from '../types';
+
+import {
+  StyledArrowWrapper,
+  StyledFlex,
+  StyledFlexItem,
+  StyledGap,
+  StyledLi,
+  StyledSelectableWrapper,
+  StyledText,
+} from './styled';
+
+interface PrivateTreeItemProps<T> {
+  state: TreeState<T>;
+  dispatch: Dispatch<Action<T>>;
+  iconless?: boolean;
+  selectable: TreeProps<T>['selectable'];
+  onExpand?: TreeProps<T>['onExpand'];
+  onCollapse?: TreeProps<T>['onCollapse'];
+  onSelect?: TreeProps<T>['onSelect'];
+}
+
+const flexItemProps: FlexItemProps = { flexShrink: 0, marginLeft: 'xxSmall' };
+
+const InternalTreeNode = <T extends unknown>({
+  children,
+  disabled,
+  dispatch,
+  icon,
+  iconless,
+  id,
+  label,
+  onCollapse,
+  onExpand,
+  onSelect,
+  selectable,
+  state,
+  value,
+}: TreeNodeProps<T> & PrivateTreeItemProps<T>): React.ReactElement<TreeNodeProps<T>> => {
+  const thisRef = useRef<TreeNodeRef<T>>({ children });
+  const nodeRef = useRef<HTMLLIElement | null>(null);
+  const selectableRef = useRef<HTMLLabelElement | null>(null);
+  const [isLoading, setIsLoading] = useState(false);
+  const expanded = useIsExpanded(state, id);
+  const selected = useIsSelected(state, value);
+
+  // Could be multiple elements in which are clicked.
+  // Typing to generic Element type since all other elements extend from it.
+  const handleNodeToggle = async (e?: React.MouseEvent<Element>) => {
+    dispatch({ type: 'FOCUS', id });
+
+    // Prevents the collapse/expand when clicking on a radio or checkbox
+    // Checks to see if every element inside the selectableRef gets clicked.
+    if (
+      (e?.target instanceof Node && selectableRef.current?.contains(e?.target)) ||
+      thisRef.current.children === undefined
+    ) {
+      return;
+    }
+
+    dispatch({ type: 'TOGGLE_NODE', id });
+
+    if (expanded) {
+      if (typeof onCollapse === 'function') {
+        const callbackValue = await onCollapse({
+          children: thisRef.current.children,
+          disabled,
+          expanded,
+          id,
+          label,
+          selected,
+          value,
+        });
+
+        if (callbackValue) {
```

Comment:
Yeah we wouldn't want a falsy value anyways. The callback value should be an object containing new children to update.

Reply:
Oh I see, I just had another look at the code. So essentially you can swap out elements of a `TreeNode` with another set of elements passed in from the outside. Can you give me an example why we want to do that? Also, after elements are swapped, would they get wiped out if a new or modified list of `nodes` get passed into `Tree`?

Reply:
Ok, in that case, I suggest we should ask the client to pass in a new list of `nodes` if it needs to display additional data. The way I see it is that `nodes` is the data source. If there are changes to the prop, the component should react to those changes. If you want to model `nodes` as the _initial_ data source, then we need to make it clear that the component won't react to any changes to the prop after its initial render. And all updates have to managed through the callbacks by manually constructing the `TreeNode`. At the moment, there are two competing ways of updating the tree (one wipes out the other). And I'd much prefer if we just stick with `nodes` instead of making `TreeNode` as a part of the public API.

Reply:
After thinking about it a bit more, I think we'll want to recommend people to use the callback method of updating the children. If this ends up being a bad for consumers we can always refactor.

I put a "notice" in the async example to use the callback method rather than updating nodes itself.

Reply:
So I went ahead and came up with something to continue along the current path (*1.*). I've renamed `nodes` to `initialNodes` and restricted re-rendering through a `usePropOnce` hook.

I do agree that we should change the component for props to passed down declaratively rather than getting set imperatively, so I went ahead and created a SPIKE ticket to explore it further.

My main reasoning behind going down approach 1 is that we already have taken into account a lot of the a11y concerns. If we go the approach where the client has to partly manage that, there's opportunities for some wrong implementations and we want to manage those misses. Not a major factor, but I also don't want to extend the release of this component a couple more weeks because we decided to change implementations. Though, I will be working with the Product team to get the spike into the next sprint or two as I feel it's important to lock it down.

Reply:
Ok cool, thank you very much @chanceaclark, I'm happy with what you're proposing 👍